### PR TITLE
Replace `production_by_klabel` by `symbols` in `KDefinition`

### PR DESCRIPTION
**Pull request by tothtamas28**
* Computes the symbol table in a single step, which is O(n) for the whole table (as opposed to O(n) per label)
* Exposes the result as a cached `FrozenDict[str, KProduction]`
* Does not filter out productions with attribute `unparseAvoid`

### CODE REVIEW

**Comment by tothtamas28** [comment]
## Profiling

```bash
$ make profile PROF_ARGS='-n0 -k test_kast_to_kore'
```

### Before

```
         134192 function calls (128318 primitive calls) in 0.134 seconds

   Ordered by: cumulative time
   List reduced from 161 to 25 due to restriction <25>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.137    0.137 /home/ttoth/git/pyk/src/pyk/konvert/_kast_to_kore.py:60(kast_to_kore)
        1    0.000    0.000    0.105    0.105 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1289(add_ksequence_under_k_productions)
        1    0.001    0.001    0.105    0.105 /home/ttoth/git/pyk/src/pyk/kast/inner.py:844(top_down)
      219    0.001    0.000    0.104    0.000 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1292(_add_ksequence_under_k_productions)
      441    0.001    0.000    0.102    0.000 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1165(production_for_klabel)
      120    0.087    0.001    0.087    0.001 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1168(<listcomp>)
  545/122    0.001    0.000    0.012    0.000 /home/ttoth/.pyenv/versions/3.10.13/lib/python3.10/functools.py:961(__get__)
        1    0.001    0.001    0.011    0.011 /home/ttoth/git/pyk/src/pyk/konvert/_kast_to_kore.py:140(_kast_to_kore)
      219    0.000    0.000    0.011    0.000 /home/ttoth/git/pyk/src/pyk/konvert/_kast_to_kore.py:158(_kinner_to_kore)
        1    0.000    0.000    0.010    0.010 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1412(add_cell_map_items)
      120    0.000    0.000    0.009    0.000 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1177(<setcomp>)
        1    0.000    0.000    0.007    0.007 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1119(cell_collection_productions)
        3    0.000    0.000    0.007    0.002 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1122(<genexpr>)
       30    0.000    0.000    0.007    0.000 /home/ttoth/git/pyk/src/pyk/konvert/_kast_to_kore.py:174(_ktoken_to_kore)
        3    0.000    0.000    0.007    0.002 /home/ttoth/git/pyk/src/pyk/prelude/string.py:13(pretty_string)
        4    0.003    0.001    0.007    0.002 {method 'join' of 'str' objects}
        3    0.000    0.000    0.007    0.002 /home/ttoth/git/pyk/src/pyk/dequote.py:14(dequote_string)
      140    0.000    0.000    0.007    0.000 /home/ttoth/git/pyk/src/pyk/kast/outer.py:870(cell_collection_productions)
     6031    0.003    0.000    0.007    0.000 /home/ttoth/.pyenv/versions/3.10.13/lib/python3.10/_collections_abc.py:828(__contains__)
      142    0.001    0.000    0.006    0.000 /home/ttoth/git/pyk/src/pyk/kast/outer.py:873(<genexpr>)
10520/5918    0.002    0.000    0.006    0.000 {built-in method builtins.hash}
        4    0.002    0.001    0.005    0.001 /home/ttoth/git/pyk/src/pyk/kast/inner.py:818(bottom_up)
        1    0.000    0.000    0.005    0.005 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1307(sort_vars)
        1    0.001    0.001    0.004    0.004 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:121(add_injections)
  813/120    0.001    0.000    0.004    0.000 /home/ttoth/git/pyk/src/pyk/utils.py:69(__hash__)
```

### After

```
         112960 function calls (112380 primitive calls) in 0.041 seconds

   Ordered by: cumulative time
   List reduced from 156 to 25 due to restriction <25>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.042    0.042 /home/ttoth/git/pyk/src/pyk/konvert/_kast_to_kore.py:60(kast_to_kore)
  546/122    0.001    0.000    0.016    0.000 /home/ttoth/.pyenv/versions/3.10.13/lib/python3.10/functools.py:961(__get__)
        1    0.001    0.001    0.011    0.011 /home/ttoth/git/pyk/src/pyk/konvert/_kast_to_kore.py:140(_kast_to_kore)
      219    0.000    0.000    0.011    0.000 /home/ttoth/git/pyk/src/pyk/konvert/_kast_to_kore.py:158(_kinner_to_kore)
        1    0.000    0.000    0.011    0.011 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1279(add_ksequence_under_k_productions)
        1    0.001    0.001    0.011    0.011 /home/ttoth/git/pyk/src/pyk/kast/inner.py:844(top_down)
        1    0.000    0.000    0.010    0.010 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1402(add_cell_map_items)
      219    0.000    0.000    0.009    0.000 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1282(_add_ksequence_under_k_productions)
        1    0.002    0.002    0.008    0.008 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1201(symbols)
        1    0.000    0.000    0.007    0.007 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1115(cell_collection_productions)
        3    0.000    0.000    0.007    0.002 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1118(<genexpr>)
       30    0.000    0.000    0.007    0.000 /home/ttoth/git/pyk/src/pyk/konvert/_kast_to_kore.py:174(_ktoken_to_kore)
      140    0.000    0.000    0.007    0.000 /home/ttoth/git/pyk/src/pyk/kast/outer.py:868(cell_collection_productions)
        3    0.000    0.000    0.007    0.002 /home/ttoth/git/pyk/src/pyk/prelude/string.py:13(pretty_string)
        4    0.003    0.001    0.007    0.002 {method 'join' of 'str' objects}
        3    0.000    0.000    0.007    0.002 /home/ttoth/git/pyk/src/pyk/dequote.py:14(dequote_string)
     5911    0.002    0.000    0.006    0.000 /home/ttoth/.pyenv/versions/3.10.13/lib/python3.10/_collections_abc.py:828(__contains__)
      142    0.001    0.000    0.006    0.000 /home/ttoth/git/pyk/src/pyk/kast/outer.py:871(<genexpr>)
        4    0.002    0.001    0.005    0.001 /home/ttoth/git/pyk/src/pyk/kast/inner.py:818(bottom_up)
        1    0.001    0.001    0.004    0.004 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:121(add_injections)
        1    0.000    0.000    0.004    0.004 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1297(sort_vars)
    27425    0.004    0.000    0.004    0.000 /home/ttoth/git/pyk/src/pyk/dequote.py:56(dequoted)
        1    0.000    0.000    0.004    0.004 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1095(productions)
     3351    0.001    0.000    0.003    0.000 /home/ttoth/git/pyk/src/pyk/kast/att.py:311(__getitem__)
     7104    0.002    0.000    0.003    0.000 /home/ttoth/git/pyk/src/pyk/utils.py:66(__getitem__)
```

File: src/pyk/kast/outer.py [hunk 1]
```diff
@@ -1224,6 +1198,22 @@ def subsorts(self, sort: KSort) -> frozenset[KSort]:
...
+        symbols: dict[str, KProduction] = {}
+        for prod in self.productions:
+            if not prod.klabel:
+                continue
+            symbol = prod.klabel.name
+            if symbol in symbols:  # Check if duplicate
+                other = symbols[symbol]
+                if prod.let(att=prod.att.drop_source()) != other.let(att=prod.att.drop_source()):
+                    prods = [other, prod]
+                    raise AssertionError(f'Found multiple productions for {symbol}: {prods}')
```
[the remark this answers is not in the document]
Comment (tothtamas28) [reply src/pyk/kast/outer.py:1212]: Good point. I've checked and `#EmtpyK` does not even seem to be present in `compiled.json`:

```python3
from pyk.kast.outer import read_kast_definition
from pyk.ktool.kompile import kompile

defn_dir = kompile('src/tests/integration/k-files/imp.k')
defn = read_kast_definition(defn_dir / 'compiled.json')
print(defn.symbols.get('#EmptyK'))
# None 
```

[the remark this answers is not in the document]
Comment (tothtamas28) [reply src/pyk/kast/outer.py:1212]: Module `KSEQ` is probably not transitively imported in the main module.

[the remark this answers is not in the document]
Comment (ehildenb) [reply src/pyk/kast/outer.py:1212]: Is it true for KEVM as well?

[the remark this answers is not in the document]
Comment (tothtamas28) [reply src/pyk/kast/outer.py:1212]: Yes, tried that as well:

```python3
from pyk.kast.outer import read_kast_definition
defn = read_kast_definition('/home/ttoth/.cache/kdist-ae73cd7/evm-semantics/llvm/compiled.json')
defn.subsorts
```

[the remark this answers is not in the document]
Comment (ehildenb) [reply src/pyk/kast/outer.py:1212]: Cool. I only see one use of `EmptyK` in KEVM, which is in `optimizer/build-spec.py`, which is probably not needed anymore! So should be good.

**Review by ehildenb — APPROVED** [review]

**Pull request merged**